File: triplate-java/src/main/java/dev/triplate/Renderer.java
package dev.triplate;

import dev.triplate.Ast.Branch;
import dev.triplate.Ast.CompiledTemplateData;
import dev.triplate.Ast.Cond;
import dev.triplate.Ast.ForNode;
import dev.triplate.Ast.IfNode;
import dev.triplate.Ast.InterpNode;
import dev.triplate.Ast.IriNode;
import dev.triplate.Ast.LangSpec;
import dev.triplate.Ast.Node;
import dev.triplate.Ast.Part;
import dev.triplate.Ast.RecordType;
import dev.triplate.Ast.ScalarKind;
import dev.triplate.Ast.ScalarType;
import dev.triplate.Ast.Schema;
import dev.triplate.Ast.TypeExpr;
import dev.triplate.Ast.ValueNode;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/** Walks the compiled body tree with a context, producing the output string. Port of {@code renderer.ts}. */
final class Renderer {
  private Renderer() {}

  private record Bound(Object value, TypeExpr type) {}

  private record Resolved(Object value, TypeExpr type, boolean present) {}

  private static final class Env {
    final Schema schema;
    final Map<String, Object> context;
    final List<Map<String, Bound>> loop;

    Env(Schema schema, Map<String, Object> context, List<Map<String, Bound>> loop) {
      this.schema = schema;
      this.context = context;
      this.loop = loop;
    }
  }

  static String render(CompiledTemplateData data, Map<String, Object> context) {
    validateContext(data.schema(), context);
    return renderNodes(data.body(), new Env(data.schema(), context, new ArrayList<>()));
  }

  // ---- context validation (up front) --------------------------------------

  private static void validateContext(Schema schema, Map<String, Object> context) {
    for (String key : context.keySet()) {
      if (!schema.byName().containsKey(key)) {
        throw new TriplateBindingError("unknown parameter: " + key);
      }
    }
    for (Ast.ParamDecl decl : schema.params()) {
      boolean present = context.containsKey(decl.name());
      if (!present) {
        if (!decl.type().optional()) {
          throw new TriplateBindingError("missing required parameter: " + decl.name());
        }
        continue;
      }
      validateValue(decl.type(), context.get(decl.name()), decl.name());
    }
  }

  private static void validateValue(TypeExpr type, Object value, String label) {
    if (type.array()) {
      if (!(value instanceof List<?> list)) {
        throw new TriplateTypeError(label + " must be a list");
      }
      if (type.min() != null && list.size() < type.min()) {
        throw new TriplateCardinalityError(
            label + " requires at least " + type.min() + " item(s), got " + list.size());
      }
      if (type.max() != null && list.size() > type.max()) {
        throw new TriplateCardinalityError(
            label + " allows at most " + type.max() + " item(s), got " + list.size());
      }
      TypeExpr elem = type.elem();
      for (int i = 0; i < list.size(); i++) {
        validateValue(elem, list.get(i), label + "[" + i + "]");
      }
    } else if (type.base() instanceof RecordType rec) {
      if (!(value instanceof Map<?, ?> obj)) {
        throw new TriplateTypeError(label + " must be an object");
      }
      for (Object key : obj.keySet()) {
        if (!rec.fields().containsKey(key)) {
          throw new TriplateBindingError("unknown field: " + label + "." + key);
        }
      }
      for (Map.Entry<String, TypeExpr> e : rec.fields().entrySet()) {
        boolean present = obj.containsKey(e.getKey());
        if (!present) {
          if (!e.getValue().optional()) {
            throw new TriplateBindingError("missing field: " + label + "." + e.getKey());
          }
          continue;
        }
        validateValue(e.getValue(), obj.get(e.getKey()), label + "." + e.getKey());
      }
    } else if (value instanceof List) {
      throw new TriplateTypeError(label + " must be a scalar, got a list");
    }
  }

  // ---- resolution ----------------------------------------------------------

  private static Resolved resolve(Env env, List<String> path, Pos pos, boolean mustExist) {
    String head = path.get(0);
    TypeExpr type;
    Object value = null;
    boolean present;
    boolean found = false;
    type = null;
    for (int i = env.loop.size() - 1; i >= 0; i--) {
      if (env.loop.get(i).containsKey(head)) {
        Bound b = env.loop.get(i).get(head);
        value = b.value();
        type = b.type();
        found = true;
        break;
      }
    }
    present = found;
    if (!found) {
      type = env.schema.byName().get(head);
      if (type == null) {
        throw new TriplateBindingError("no binding for variable: " + head, pos.line(), pos.column());
      }
      present = env.context.containsKey(head);
      value = present ? env.context.get(head) : null;
    }
    for (int i = 1; i < path.size(); i++) {
      String field = path.get(i);
      if (!(type.base() instanceof RecordType rec)) {
        throw new TriplateTypeError(String.join(".", path.subList(0, i)) + " is not an object",
            pos.line(), pos.column());
      }
      TypeExpr ft = rec.fields().get(field);
      if (ft == null) {
        throw new TriplateBindingError("unknown field: " + String.join(".", path.subList(0, i + 1)),
            pos.line(), pos.column());
      }
      if (present && value instanceof Map<?, ?> map && map.containsKey(field)) {
        value = map.get(field);
      } else {
        present = false;
        value = null;
      }
      type = ft;
    }
    if (mustExist && !present) {
      throw new TriplateBindingError(
          "no value for " + String.join(".", path) + " (an optional value? guard it with {% if %})",
          pos.line(), pos.column());
    }
    return new Resolved(value, type, present);
  }

  private static ScalarType scalarOf(TypeExpr type, List<String> path, Pos pos) {
    if (type.array()) {
      throw new TriplateTypeError(String.join(".", path) + " is a list; loop over it with {% for %}",
          pos.line(), pos.column());
    }
    if (type.base() instanceof RecordType) {
      throw new TriplateTypeError(String.join(".", path) + " is an object; reference a field",
          pos.line(), pos.column());
    }
    return (ScalarType) type.base();
  }

  // ---- rendering -----------------------------------------------------------

  private static String renderNodes(List<Node> nodes, Env env) {
    StringBuilder out = new StringBuilder();
    for (Node node : nodes) {
      if (node instanceof Ast.TextNode t) {
        out.append(t.value());
      } else if (node instanceof ValueNode v) {
        Pos pos = Pos.at(v.line(), v.column());
        Resolved r = resolve(env, v.path(), pos, true);
        out.append(serializeScalar(scalarOf(r.type(), v.path(), pos), r.value(), pos));
      } else if (node instanceof InterpNode i) {
        out.append(renderInterp(env, i));
      } else if (node instanceof IriNode ir) {
        out.append(renderIri(env, ir));
      } else if (node instanceof ForNode f) {
        out.append(renderFor(env, f));
      } else if (node instanceof IfNode iff) {
        out.append(renderIf(env, iff));
      }
    }
    return out.toString();
  }

  private static String serializeScalar(ScalarType t, Object value, Pos pos) {
    return switch (t.kind()) {
      case IRI -> Serializers.serializeIri(value, pos);
      case PNAME -> Serializers.serializePname(value, pos);
      case STRING -> Serializers.serializeString(value, pos);
      case INT -> Serializers.serializeInt(value, pos);
      case DECIMAL -> Serializers.serializeDecimal(value, pos);
      case DOUBLE -> Serializers.serializeDouble(value, pos);
      case BOOL -> Serializers.serializeBoolean(value, pos);
      case DATE -> Serializers.serializeDate(value, pos);
      case DATE_TIME -> Serializers.serializeDateTime(value, pos);
      case TIME -> Serializers.serializeTime(value, pos);
      case LITERAL -> Serializers.serializeTypedLiteral(value, t.datatype(), pos);
      case TERM -> Serializers.serializeTerm(value, pos);
      case RAW -> Serializers.serializeRaw(value, pos);
      case CUSTOM -> {
        TypeRegistry.CustomSerializer fn = TypeRegistry.get(t.name());
        if (fn == null) {
          throw new TriplateTypeError("unknown custom type: " + t.name(), pos.line(), pos.column());
        }
        yield fn.serialize(value, pos.line(), pos.column());
      }
    };
  }

  private record HoleText(boolean raw, String text) {}

  /** Lexical form of a value inside a string/IRI construct. {@code raw} = insert verbatim. */
  private static HoleText holeLexical(Env env, Ast.HolePart part) {
    Pos pos = Pos.at(part.line(), part.column());
    Resolved r = resolve(env, part.path(), pos, true);
    ScalarType t = scalarOf(r.type(), part.path(), pos);
    Object value = r.value();
    return switch (t.kind()) {
      case RAW -> new HoleText(true, Serializers.serializeRaw(value, pos));
      case STRING -> new HoleText(false, Serializers.requireString(value, "string", pos));
      case IRI -> new HoleText(false, Serializers.validateIri(value, pos));
      case PNAME -> new HoleText(false, Serializers.serializePname(value, pos));
      case INT -> new HoleText(false, Serializers.serializeInt(value, pos));
      case DECIMAL -> new HoleText(false, Serializers.serializeDecimal(value, pos));
      case DOUBLE -> new HoleText(false, Serializers.serializeDouble(value, pos));
      case BOOL -> new HoleText(false, Serializers.serializeBoolean(value, pos));
      default -> throw new TriplateTypeError(
          "type " + t.kind() + " cannot be interpolated; use it as a standalone ${ … }",
          part.line(), part.column());
    };
  }

  private static String resolveLang(Env env, LangSpec lang, Pos pos) {
    if (lang instanceof Ast.StaticLang sl) {
      return Serializers.validateLangTag(sl.tag(), pos);
    }
    Ast.PathLang pl = (Ast.PathLang) lang;
    Resolved r = resolve(env, pl.path(), pos, true);
    return Serializers.validateLangTag(r.value(), pos);
  }

  private static String renderInterp(Env env, InterpNode node) {
    StringBuilder content = new StringBuilder();
    for (Part part : node.parts()) {
      if (part instanceof Ast.TextPart tp) {
        content.append(Serializers.escapeString(tp.text()));
        continue;
      }
      HoleText h = holeLexical(env, (Ast.HolePart) part);
      content.append(h.raw() ? h.text() : Serializers.escapeString(h.text()));
    }
    StringBuilder out = new StringBuilder("\"").append(content).append("\"");
    Pos pos = Pos.at(node.line(), node.column());
    if (node.lang() != null) out.append("@").append(resolveLang(env, node.lang(), pos));
    else if (node.datatype() != null) out.append("^^").append(node.datatype());
    return out.toString();
  }

  private static String renderIri(Env env, IriNode node) {
    StringBuilder body = new StringBuilder();
    for (Part part : node.parts()) {
      if (part instanceof Ast.TextPart tp) {
        body.append(tp.text());
        continue;
      }
      HoleText h = holeLexical(env, (Ast.HolePart) part);
      body.append(h.raw() ? h.text() : Serializers.percentEncode(h.text()));
    }
    Pos pos = Pos.at(node.line(), node.column());
    return "<" + Serializers.validateAssembledIri(body.toString(), pos) + ">";
  }

  private static String renderFor(Env env, ForNode node) {
    Pos pos = Pos.at(node.line, node.column);
    Resolved r = resolve(env, node.source, pos, true);
    if (!r.type().array() || !(r.value() instanceof List<?> list)) {
      throw new TriplateTypeError(String.join(".", node.source) + " is not a list", node.line, node.column);
    }
    TypeExpr elemType = r.type().elem();
    List<String> chunks = new ArrayList<>();
    for (Object item : list) {
      List<Map<String, Bound>> loop = new ArrayList<>(env.loop);
      Map<String, Bound> frame = new HashMap<>();
      frame.put(node.item, new Bound(item, elemType));
      loop.add(frame);
      chunks.add(renderNodes(node.body, new Env(env.schema, env.context, loop)));
    }
    if (node.join == null) return String.join("", chunks);
    String separator;
    if (node.joinExact) {
      separator = node.join;
    } else {
      String trimmed = node.join.strip();
      separator = trimmed.isEmpty() ? " " : " " + trimmed + " ";
    }
    if (chunks.isEmpty()) return "";
    String out = chunks.get(0);
    for (int i = 1; i < chunks.size(); i++) {
      out = trimEnd(out) + separator + trimStart(chunks.get(i));
    }
    return out;
  }

  private static String trimEnd(String s) {
    return s.replaceAll("\\s+$", "");
  }

  private static String trimStart(String s) {
    return s.replaceFirst("^\\s+", "");
  }

  private static boolean evalCond(Env env, Cond cond) {
    Pos pos = Pos.at(cond.line(), cond.column());
    Resolved r = resolve(env, cond.path(), pos, false);
    boolean result;
    if (r.type().array()) {
      result = r.present() && r.value() instanceof List<?> list && !list.isEmpty();
    } else if (r.type().base() instanceof ScalarType st && st.kind() == ScalarKind.BOOL) {
      result = r.present() && Boolean.TRUE.equals(r.value());
    } else if (r.type().optional()) {
      result = r.present();
    } else {
      String what = r.type().base() instanceof RecordType
          ? "object"
          : ((ScalarType) r.type().base()).kind().toString().toLowerCase();
      throw new TriplateTypeError(
          "condition on required " + what + " '" + String.join(".", cond.path()) + "' is always true",
          cond.line(), cond.column());
    }
    return cond.negated() ? !result : result;
  }

  private static String renderIf(Env env, IfNode node) {
    for (Branch branch : node.branches) {
      if (evalCond(env, branch.cond)) return renderNodes(branch.body, env);
    }
    if (node.elseBody != null) return renderNodes(node.elseBody, env);
    return "";
  }
}

File: triplate-py/src/triplate/_renderer.py
"""AST renderer + context validation. Mirrors renderer.ts."""

import re

from ._ast import (
    ForNode,
    IfNode,
    InterpNode,
    IriNode,
    LangStatic,
    PartHole,
    TextNode,
    TypeExpr,
    ValueNode,
)
from ._registry import get_custom_type
from ._serializers import (
    escape_string,
    percent_encode,
    require_string,
    serialize_bool,
    serialize_date,
    serialize_date_time,
    serialize_decimal,
    serialize_double,
    serialize_int,
    serialize_iri,
    serialize_pname,
    serialize_raw,
    serialize_string,
    serialize_term,
    serialize_time,
    serialize_typed_literal,
    validate_assembled_iri,
    validate_iri,
    validate_lang_tag,
)
from .errors import TriplateBindingError, TriplateCardinalityError, TriplateTypeError

_TRAILING_WS = re.compile(r"\s+$")
_LEADING_WS = re.compile(r"^\s+")


def render(data, context):
    _validate_context(data.schema, context)
    env = {"schema": data.schema, "context": context, "loop": []}
    return _render(data.body, env)


# ---- context validation --------------------------------------------------

def _validate_context(schema, context):
    for key in context:
        if key not in schema.by_name:
            raise TriplateBindingError(f"unknown parameter: {key}")
    for decl in schema.params:
        present = decl.name in context
        if not present:
            if not decl.type.optional:
                raise TriplateBindingError(f"missing required parameter: {decl.name}")
            continue
        _validate_value(decl.type, context[decl.name], decl.name)


def _validate_value(t, value, label):
    if t.array:
        if not isinstance(value, (list, tuple)):
            raise TriplateTypeError(f"{label} must be a list")
        if t.min is not None and len(value) < t.min:
            raise TriplateCardinalityError(f"{label} requires at least {t.min} item(s), got {len(value)}")
        if t.max is not None and len(value) > t.max:
            raise TriplateCardinalityError(f"{label} allows at most {t.max} item(s), got {len(value)}")
        elem = TypeExpr(t.base, False, False, None, None)
        for i, el in enumerate(value):
            _validate_value(elem, el, f"{label}[{i}]")
    elif t.base["kind"] == "record":
        if not isinstance(value, dict):
            raise TriplateTypeError(f"{label} must be an object")
        for key in value:
            if key not in t.base["fields"]:
                raise TriplateBindingError(f"unknown field: {label}.{key}")
        for fname, ftype in t.base["fields"].items():
            if fname not in value:
                if not ftype.optional:
                    raise TriplateBindingError(f"missing field: {label}.{fname}")
                continue
            _validate_value(ftype, value[fname], f"{label}.{fname}")
    elif isinstance(value, (list, tuple)):
        raise TriplateTypeError(f"{label} must be a scalar, got a list")


# ---- resolution ----------------------------------------------------------

def _resolve(env, path, pos, must_exist):
    head = path[0]
    value = None
    type_ = None
    present = True
    found = False
    for scope in reversed(env["loop"]):
        if head in scope:
            value, type_ = scope[head]
            found = True
            break
    if not found:
        type_ = env["schema"].by_name.get(head)
        if type_ is None:
            raise TriplateBindingError(f"no binding for variable: {head}", pos.line, pos.column)
        present = head in env["context"]
        value = env["context"].get(head)
    for i in range(1, len(path)):
        field = path[i]
        if type_.base["kind"] != "record":
            raise TriplateTypeError(f"{'.'.join(path[:i])} is not an object", pos.line, pos.column)
        ft = type_.base["fields"].get(field)
        if ft is None:
            raise TriplateBindingError(f"unknown field: {'.'.join(path[:i + 1])}", pos.line, pos.column)
        if present and isinstance(value, dict) and field in value:
            value = value[field]
        else:
            present = False
            value = None
        type_ = ft
    if must_exist and not present:
        raise TriplateBindingError(
            f"no value for {'.'.join(path)} (an optional value? guard it with {{% if %}})",
            pos.line,
            pos.column,
        )
    return value, type_, present


def _scalar_of(t, path, pos):
    if t.array:
        raise TriplateTypeError(f"{'.'.join(path)} is a list; loop over it with {{% for %}}", pos.line, pos.column)
    if t.base["kind"] == "record":
        raise TriplateTypeError(f"{'.'.join(path)} is an object; reference a field", pos.line, pos.column)
    return t.base


# ---- rendering -----------------------------------------------------------

def _render(nodes, env):
    out = []
    for node in nodes:
        if isinstance(node, TextNode):
            out.append(node.value)
        elif isinstance(node, ValueNode):
            value, type_, _ = _resolve(env, node.path, node, True)
            out.append(_serialize_scalar(_scalar_of(type_, node.path, node), value, node))
        elif isinstance(node, InterpNode):
            out.append(_render_interp(env, node))
        elif isinstance(node, IriNode):
            out.append(_render_iri(env, node))
        elif isinstance(node, ForNode):
            out.append(_render_for(env, node))
        elif isinstance(node, IfNode):
            out.append(_render_if(env, node))
    return "".join(out)


def _serialize_scalar(t, value, pos):
    k = t["kind"]
    if k == "iri":
        return serialize_iri(value, pos)
    if k == "pname":
        return serialize_pname(value, pos)
    if k == "string":
        return serialize_string(value, pos)
    if k == "int":
        return serialize_int(value, pos)
    if k == "decimal":
        return serialize_decimal(value, pos)
    if k == "double":
        return serialize_double(value, pos)
    if k == "bool":
        return serialize_bool(value, pos)
    if k == "date":
        return serialize_date(value, pos)
    if k == "dateTime":
        return serialize_date_time(value, pos)
    if k == "time":
        return serialize_time(value, pos)
    if k == "literal":
        return serialize_typed_literal(value, t["datatype"], pos)
    if k == "term":
        return serialize_term(value, pos)
    if k == "raw":
        return serialize_raw(value, pos)
    if k == "custom":
        fn = get_custom_type(t["name"])
        if fn is None:
            raise TriplateTypeError(f"unknown custom type: {t['name']}", pos.line, pos.column)
        return fn(value, pos)
    raise TriplateTypeError(f"unknown type kind: {k}", pos.line, pos.column)


def _hole_lexical(env, part):
    value, type_, _ = _resolve(env, part.path, part, True)
    t = _scalar_of(type_, part.path, part)
    k = t["kind"]
    if k == "raw":
        return True, serialize_raw(value, part)
    if k == "string":
        return False, require_string(value, "string", part)
    if k == "iri":
        return False, validate_iri(value, part)
    if k == "pname":
        return False, serialize_pname(value, part)
    if k == "int":
        return False, serialize_int(value, part)
    if k == "decimal":
        return False, serialize_decimal(value, part)
    if k == "double":
        return False, serialize_double(value, part)
    if k == "bool":
        return False, serialize_bool(value, part)
    raise TriplateTypeError(
        f"type {k} cannot be interpolated; use it as a standalone ${{ … }}", part.line, part.column
    )


def _resolve_lang(env, lang, pos):
    if isinstance(lang, LangStatic):
        return validate_lang_tag(lang.static, pos)
    value, _, _ = _resolve(env, lang.path, pos, True)
    return validate_lang_tag(value, pos)


def _render_interp(env, node):
    content = []
    for part in node.parts:
        if isinstance(part, PartHole):
            raw, text = _hole_lexical(env, part)
            content.append(text if raw else escape_string(text))
        else:
            content.append(escape_string(part.text))
    out = '"' + "".join(content) + '"'
    if node.lang is not None:
        out += "@" + _resolve_lang(env, node.lang, node)
    elif node.datatype is not None:
        out += "^^" + node.datatype
    return out


def _render_iri(env, node):
    body = []
    for part in node.parts:
        if isinstance(part, PartHole):
            raw, text = _hole_lexical(env, part)
            body.append(text if raw else percent_encode(text))
        else:
            body.append(part.text)
    return "<" + validate_assembled_iri("".join(body), node) + ">"


def _render_for(env, node):
    value, type_, _ = _resolve(env, node.source, node, True)
    if not type_.array or not isinstance(value, (list, tuple)):
        raise TriplateTypeError(f"{'.'.join(node.source)} is not a list", node.line, node.column)
    elem = TypeExpr(type_.base, False, False, None, None)
    chunks = []
    for item in value:
        child = {"schema": env["schema"], "context": env["context"], "loop": env["loop"] + [{node.item: (item, elem)}]}
        chunks.append(_render(node.body, child))
    if node.join is None:
        return "".join(chunks)
    if node.join_exact:
        separator = node.join
    else:
        trimmed = node.join.strip()
        separator = f" {trimmed} " if trimmed else " "
    out = chunks[0] if chunks else ""
    for chunk in chunks[1:]:
        out = _TRAILING_WS.sub("", out) + separator + _LEADING_WS.sub("", chunk)
    return out


def _eval_cond(env, cond):
    value, type_, present = _resolve(env, cond.path, cond, False)
    if type_.array:
        result = present and isinstance(value, (list, tuple)) and len(value) > 0
    elif type_.base["kind"] == "bool":
        result = present and value is True
    elif type_.optional:
        result = present
    else:
        what = "object" if type_.base["kind"] == "record" else type_.base["kind"]
        raise TriplateTypeError(
            f"condition on required {what} '{'.'.join(cond.path)}' is always true",
            cond.line,
            cond.column,
        )
    return (not result) if cond.negated else result


def _render_if(env, node):
    for branch in node.branches:
        if _eval_cond(env, branch.cond):
            return _render(branch.body, env)
    if node.else_body is not None:
        return _render(node.else_body, env)
    return ""
